pandas/core/base.py

```diff
@@ -954,19 +955,30 @@ def value_counts(self, normalize=False, sort=True, ascending=False,
 
     def unique(self):
         """
-        Return array of unique values in the object. Significantly faster than
+        Return Index of unique values in the object. Significantly faster than
         numpy.unique. Includes NA values.
 
         Returns
         -------
-        uniques : ndarray
+        uniques : Index
         """
-        from pandas.core.nanops import unique1d
-        values = self.values
-        if hasattr(values, 'unique'):
-            return values.unique()
+        values = self._values
 
-        return unique1d(values)
+        if hasattr(values, 'unique'):
+            result = values.unique()
+        else:
+            from pandas.core.nanops import unique1d
+            result = unique1d(values)
+
+        if isinstance(self, ABCCategoricalIndex):
+            # CategoricalIndex._shallow_copy uses keeps original categories
```

Comment:
We have to update categories because ``Categorical.unique`` reorderes categories if its ``ordered=False``. This can be simplified if we change ``Categorical.unique`` to keep ``categories``.

- https://github.com/pydata/pandas/blob/master/pandas/core/categorical.py#L1753